Convert SharedDatabaseConnect Dialog to Javafx

Siedlerchr opened the pull request:
Connection works so far.  No more freezes regarding addTabs. in basePanel
Somehow the dialog looks odd, @tobiasdiez  Any idea what is missing? The checkboxes are not visible correct. I just see that it looks really odd in he maintable-beta branch, too. e..g. int he Event Log

Edit// The conection and the synchronization seems to be working.  

Fixes: #3419   PostgresSQL database connection now supports SSL

TODO for documentaition:
Create and configure Server certs
https://www.postgresql.org/docs/current/static/ssl-tcp.html

Convert to keystore format
https://jdbc.postgresql.org/documentation/head/ssl-client.html

![grafik](https://user-images.githubusercontent.com/320228/44865756-7cd7a280-ac83-11e8-99d0-9c619a929163.png)

Currently I only tested it with PostgresSQL, but I see that mySQL should also work
https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-using-ssl.html



<!-- describe the changes you have made here: what, why, ... 
     Link issues by using the following pattern: [#333](https://github.com/JabRef/jabref/issues/333) or [koppor#49](https://github.com/koppor/jabref/issues/47).
     The title of the PR must not reference an issue, because GitHub does not support autolinking there. -->


----

TODO: 
- check conflict merge dialog


- [x] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [x] Manually tested changed features in running JabRef
- [x] Screenshots added in PR description (for bigger UI changes)
- [x] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

tobiasdiez commented:
The display problem probably results from us playing around with different styles and changing defaults. Since there is no checkbox in the main view, it was not noticed that some of the input controls look strange. I'll have a look at it later...

Siedlerchr commented:
The database connection works so far. Tested with PostgresSQL. I reworked the dialog and fixed an error in the listener which led to freezes. 
The MergeEntryDialog is not yet converted, this will be done in a future PR.

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialog.fxml:
```diff
@@ -0,0 +1,74 @@
+                        <Label text="%Database Type:" />
+                        <ComboBox fx:id="cmbDbType" prefHeight="25.0" prefWidth="606.0" GridPane.columnIndex="1" GridPane.columnSpan="2" />
+                        <Label text="%Host/Port:" GridPane.rowIndex="1" />
+                        <TextField fx:id="tbHost" GridPane.columnIndex="1" GridPane.rowIndex="1" />
```
I don't have a strong opinion on this, but this Hungarian notation is against the usual naming convention. Probably, `host` or `hostTextField` is preferable. @koppor what's your opinion about this?

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,320 @@
+        this.frame = frame;
+        this.setTitle(Localization.lang("Connect to shared database"));
+
+        this.getDialogPane().getButtonTypes().addAll(connect, ButtonType.CANCEL);
```
It is preferable to set the buttons in the fxml
https://github.com/JabRef/jabref/blob/390c38740997201c6ab6c5e66bf53c2991507a48/src/main/java/org/jabref/gui/help/AboutDialog.fxml#L99-L100
and then add the action in the code-behind
https://github.com/JabRef/jabref/blob/390c38740997201c6ab6c5e66bf53c2991507a48/src/main/java/org/jabref/gui/help/AboutDialogView.java#L36

Siedlerchr replied:
Unfortunately this results in an NPE which I don't understand.  The lookUpButton returns null.

tobiasdiez replied:
Did you also tried it without the surrounding `buttontypes` tag?

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,320 @@
+
+    }
+
+    private void openDatabase(ActionEvent e) {
```
This method and essentially everything below should be moved to the `ViewModel`. The view should contain no logic whatsoever and just binds the fxml file to the properties in the view model.

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,320 @@
+        cmbDbType.getSelectionModel().select(0);
+        viewModel.selectedDbmstypeProperty().bind(cmbDbType.getSelectionModel().selectedItemProperty());
+
+        viewModel.selectedDbmstypeProperty().addListener((observable, oldValue, newValue) -> {
```
`EasyBind.subscribe` is usually easier to read. Moreover, this logic should also be moved to the view model.

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,320 @@
+        FileDialogConfiguration fileDialogConfiguration = new FileDialogConfiguration.Builder()
+                                                                                               .addExtensionFilter(FileType.BIBTEX_DB)
+                                                                                               .withDefaultExtension(FileType.BIBTEX_DB)
+                                                                                               .withInitialDirectory(Globals.prefs.get(JabRefPreferences.WORKING_DIRECTORY))
```
Please extract the `Globals.prefs` dependency. The new JavaFX stuff should no longer depend on `Globals`.

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,320 @@
+                }
+            }
+
+            return; // setLoadingConnectButtonText(false) should not be reached regularly.
```
what is the comment trying to tell us?

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,320 @@
+        } catch (DatabaseNotSupportedException exception) {
+            ButtonType openHelp = new ButtonType("Open Help", ButtonData.OTHER);
+
+            Optional<ButtonType> result = dialogService.showCustomButtonDialogAndWait(AlertType.INFORMATION,
```
Information or Warning?

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,320 @@
+    }
+
+    private void setLoadingConnectButtonText(boolean isLoading) {
+        btnConnect.setDisable(isLoading);
```
The `viewModel` should have a boolean property `connectionInProgress` so that the functionality of this method can be realized by binding.

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,320 @@
+    private DBMSConnectionProperties connectionProperties;
+    private SharedDatabaseLoginDialogViewModel viewModel;
+
+    public SharedDatabaseLoginDialogView(JabRefFrame frame) {
```
I really don't like that we need a frame here. Can you please investigate if it can be replaced by the `StateManager` (maybe after extending the latter).

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,320 @@
+    }
+
+    private void checkFields() throws JabRefException {
+        if (isEmptyField(tbHost)) {
```
Validation should be done using the validation framework so that the user get immediate feedback (have a look at the usage of ControlsFxVisualizer).

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseUIManager.java:
```diff
@@ -37,32 +38,40 @@
+                jabRefFrame.closeCurrentTab();
+                new SharedDatabaseLoginDialogView(jabRefFrame).showAndWait();
+
+            }
```
`else if` (and remove empty line above)

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseUIManager.java:
```diff
@@ -86,12 +95,12 @@ public void listen(SharedEntryNotPresentEvent event) {
 
-            SwingUtilities.invokeLater(() -> panel.closeBottomPane());
+            dialogService.showInformationDialogAndWait(Localization.lang("Shared entry is no longer present"),
+                                                       Localization.lang("The BibEntry you currently work on has been deleted on the shared side.")
```
`BibEntry` is for our internal usage and `entry` should be used for dialogs.

Siedlerchr commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialog.fxml:
```diff
@@ -0,0 +1,81 @@
+   </header>
+   <buttonTypes>
+      <ButtonType fx:constant="CLOSE" />
+      <ButtonType fx:id="connectButton" text="Connect" />
```
Here I add the button

Siedlerchr commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,320 @@
+    @FXML private TextField tbFolder;
+    @FXML private Button btnBrowse;
+    @FXML private CheckBox chkAutosave;
+    @FXML private ButtonType connectButton;
```
The buttonType is initalized here

Siedlerchr commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,320 @@
+                  .load()
+                  .setAsContent(this.getDialogPane());
+
+        ControlHelper.setAction(connectButton, this.getDialogPane(), event -> openDatabase());
```
In this method the lookup method from the dialogPane returns null. I don't understand why.

Siedlerchr replied:
Somehow the buttons are in the content property of the dialog pane. Does not make any sense to me

Siedlerchr commented:
Yes I tried both with and without the tags. Same behavior. I don't
understand this.

Tobias Diez <notifications@github.com> schrieb am Fr., 25. Mai 2018, 15:39:

> *@tobiasdiez* commented on this pull request.
> ------------------------------
>
> In src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java
> <https://github.com/JabRef/jabref/pull/4040#discussion_r190896614>:
>
> > +    @Inject private DialogService dialogService;
> +
> +    private final SharedDatabasePreferences prefs = new SharedDatabasePreferences();
> +
> +    private final ButtonType connect = new ButtonType("Connect", ButtonData.YES);
> +    private final Button btnConnect;
> +    private final JabRefFrame frame;
> +
> +    private DBMSConnectionProperties connectionProperties;
> +    private SharedDatabaseLoginDialogViewModel viewModel;
> +
> +    public SharedDatabaseLoginDialogView(JabRefFrame frame) {
> +        this.frame = frame;
> +        this.setTitle(Localization.lang("Connect to shared database"));
> +
> +        this.getDialogPane().getButtonTypes().addAll(connect, ButtonType.CANCEL);
>
> Did you also tried it without the surrounding buttontypes tag?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JabRef/jabref/pull/4040#discussion_r190896614>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AATi5GBJ--hYtjBd1unf_oiZTC9_5v_Zks5t2AmsgaJpZM4UGDLC>
> .
>

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,320 @@
+
+        ViewLoader.view(this)
+                  .load()
+                  .setAsContent(this.getDialogPane());
```
Should be `.setAsDialogPane(this);`

Siedlerchr commented:
Now moved all stuff and implemented the validationto the viewmodel and will tackle the SharedUiManager stuff next

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,120 @@
+        tbFolder.disableProperty().bind(chkAutosave.selectedProperty().not());
+        chkAutosave.selectedProperty().bindBidirectional(viewModel.autosaveProperty());
+
+        Platform.runLater(() -> {
```
Please add a comment that this is a workaround and add a link to the issue in controlsfx

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,120 @@
+        ControlHelper.setAction(connectButton, this.getDialogPane(), event -> openDatabase());
+        btnConnect = (Button) this.getDialogPane().lookupButton(connectButton);
+        setLoadingConnectButtonText();
+        btnConnect.disableProperty().bind(viewModel.formValidation().validProperty().not());
```
move the last line to `initialize`

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogView.java:
```diff
@@ -0,0 +1,120 @@
+
+    private void setLoadingConnectButtonText() {
+
+        if (viewModel.loadingProperty().get()) {
```
This should be done via binding, because otherwise setting `loading = true` does not change the text to "Connecting...". You can use `EasyBind.map` to convert the boolean to the appropriate text and then bind this text to `btnConnect.textProperty`

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogViewModel.java:
```diff
@@ -0,0 +1,324 @@
+    private final ListProperty<DBMSType> allDBMSTypes = new SimpleListProperty<>(FXCollections.observableArrayList(DBMSType.values()));
+    private final ObjectProperty<DBMSType> selectedDBMSType = new SimpleObjectProperty<>(allDBMSTypes.get(0));
+
+    private final StringProperty db = new SimpleStringProperty("");
```
no abbreviation here and above (selectedDatabaseType).

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogViewModel.java:
```diff
@@ -0,0 +1,324 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SharedDatabaseLoginDialogViewModel.class);
+
+    private final ListProperty<DBMSType> allDBMSTypes = new SimpleListProperty<>(FXCollections.observableArrayList(DBMSType.values()));
```
since this is an static enum, you probably don't need to wrap it in a list property.

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogViewModel.java:
```diff
@@ -0,0 +1,324 @@
+    private final DialogService dialogService;
+    private final SharedDatabasePreferences prefs = new SharedDatabasePreferences();
+
+    private final Validator dbvalidator;
```
no abbreviation

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogViewModel.java:
```diff
@@ -0,0 +1,324 @@
+            port.setValue(Integer.toString(selected.getDefaultPort()));
+        });
+
+        Predicate<String> predicate = input -> (input != null) && !input.trim().isEmpty();
```
StringUtil has a `isNotBlank` method that does exactly this.

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogViewModel.java:
```diff
@@ -0,0 +1,324 @@
+    private final StringProperty port = new SimpleStringProperty("");
+    private final StringProperty user = new SimpleStringProperty("");
+    private final StringProperty password = new SimpleStringProperty("");
+    private final StringProperty folder = new SimpleStringProperty("");
```
is it a folder or actually a file?

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogViewModel.java:
```diff
@@ -0,0 +1,324 @@
+        hostvalidator = new FunctionBasedValidator<>(host, predicate, ValidationMessage.error(Localization.lang("Required field \"%0\" is empty.", Localization.lang("Port"))));
+        portValidator = new FunctionBasedValidator<>(port, predicate, ValidationMessage.error(Localization.lang("Required field \"%0\" is empty.", Localization.lang("Host"))));
+        uservalidator = new FunctionBasedValidator<>(user, predicate, ValidationMessage.error(Localization.lang("Required field \"%0\" is empty.", Localization.lang("User"))));
+        folderValidator = new FunctionBasedValidator<>(folder, predicate, ValidationMessage.error(Localization.lang("Please enter a valid file path.")));
```
I think we should use the nio-path methods to check if the provided path is actually valid.

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogViewModel.java:
```diff
@@ -0,0 +1,324 @@
+
+    public void openDatabase() {
+
+        connectionProperties = new DBMSConnectionProperties();
```
I would convert this to a local variable and pass it to `openSharedDatabase` as a argument (because it may be null and a caller of `openSharedDatabase` has to know that `connectionProperties` need to be init first...)

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogViewModel.java:
```diff
@@ -0,0 +1,324 @@
+                                                                                           Localization.lang("Overwrite file"),
+                                                                                           Localization.lang("Cancel"));
+                if (!overwriteFilePressed) {
+                    // tbFolder.requestFocus();
```
remove comment

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogViewModel.java:
```diff
@@ -0,0 +1,324 @@
+    /**
+     * Fetches possibly saved data and configures the control elements respectively.
+     */
+    public void applyPreferences() {
```
Make this private and call from constructor?

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogViewModel.java:
```diff
@@ -0,0 +1,324 @@
+
+    private boolean isSharedDatabaseAlreadyPresent() {
+        List<BasePanel> panels = frame.getBasePanelList();
+        return panels.parallelStream().anyMatch(panel -> {
```
Can you try to move this check using `frame` to `StateManager` (it also gives access to the list of open databases).

Siedlerchr replied:
It currently only support the active database, needs probably be extended in JabRefFrame, there is this line:   EasyBind.subscribe(tabbedPane.getSelectionModel().selectedItemProperty(), e -> {

tobiasdiez commented on src/main/java/org/jabref/gui/shared/SharedDatabaseLoginDialogViewModel.java:
```diff
@@ -0,0 +1,359 @@
+    }
+
+    private void setPreferences() {
+        prefs.setType(selectedDBMSType.getValue().toString());
```
Return the preferences instead of setting a local variable? Makes the flow more logical in my opinion.

tobiasdiez commented on src/main/java/org/jabref/logic/shared/DBMSConnection.java:
```diff
@@ -21,23 +22,30 @@
-            this.connection = DriverManager.getConnection(
-                    properties.getType().getUrl(properties.getHost(), properties.getPort(), properties.getDatabase()),
-                    properties.getUser(), properties.getPassword());
+            Properties props = new Properties();
```
Can you please move these additions to the `DBMSConnectionProperties ` class as two methods `Properties  asJavaProperties()` and `String getUrl`.

tobiasdiez commented on src/main/java/org/jabref/logic/shared/DBMSConnectionProperties.java:
```diff
@@ -100,15 +101,23 @@ public void setPassword(String password) {
+               && this.host.equalsIgnoreCase(properties.getHost())
+               && (this.port == properties.getPort())
+               && this.database.equals(properties.getDatabase())
+               && this.user.equals(properties.getUser());
```
add `useSSL` in equals and hash

tobiasdiez commented on src/main/java/org/jabref/logic/shared/PostgreSQLProcessor.java:
```diff
@@ -89,7 +89,8 @@ public void startNotificationListener(DBMSSynchronizer dbmsSynchronizer) {
             PGConnection pgConnection = connection.unwrap(PGConnection.class);
             listener = new PostgresSQLNotificationListener(dbmsSynchronizer, pgConnection);
-            listener.start();
+            new Thread(listener, "PostgresSQLNotificationListener").start();
```
Better reuse our threading infrastructure so that we are sure that the thread is shutdown on exit etc.

tobiasdiez commented on src/main/java/org/jabref/logic/shared/listener/PostgresSQLNotificationListener.java:
```diff
@@ -13,23 +13,29 @@
 
+    public void start() {
+        stop = false;
+        new Thread(this).start();
```
same

tobiasdiez commented on src/main/java/org/jabref/logic/shared/prefs/SharedDatabasePreferences.java:
```diff
@@ -71,6 +72,11 @@ public boolean getRememberPassword() {
     }
 
+    public boolean isUseSSL()
+    {
```
code style
